Fixes #9237 - Decouple QTP idleTimeout from pool shrink rate.

sbordet opened the pull request:
Introduced `QueuedThreadPool.idleTimeoutMaxShrinkCount` to be the number of idle threads that are exited in one idle timeout.

When set to 1 (the default), the old behavior is reproduced: expiring 1 thread every idle timeout. When set to larger values, allows to keep around the threads for the idle timeout (in case of further load spikes), but allows to quickly recover OS memory when they are truly idle.

For example, with 2000 threads, 30 seconds idle timeout and idleTimeoutMaxShrinkCount=1, it will take 995 minutes (about 16.5 hrs) to shrink the pool back to 10 threads. By setting idleTimeoutMaxShrinkCount=100, the thread pool can be shrunk to 10 threads in about 10 minutes.

sbordet committed:
Fixes #9237 - Decouple QTP idleTimeout from pool shrink rate.

Introduced `QueuedThreadPool.idleTimeoutMaxShrinkCount` to be the number of idle threads that are exited in one idle timeout.

When set to 1 (the default), the old behavior is reproduced: expiring 1 thread every idle timeout.
When set to larger values, allows to keep around the threads for the idle timeout (in case of further load spikes), but allows to quickly recover OS memory when they are truly idle.

For example, with 2000 threads, 30 seconds idle timeout and idleTimeoutMaxShrinkCount=1, it will take 995 minutes (about 16.5 hrs) to shrink the pool back to 10 threads.
By setting idleTimeoutMaxShrinkCount=100, the thread pool can be shrunk to 10 threads in about 10 minutes.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
Alternative to #9354.

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -1027,54 +1057,68 @@ public void run()
+
+                                    // Add shrink periods until the window is full.
+                                    long nextShrink = prevShrink + shrinkPeriod;
+                                    if (NanoTime.isBefore(now, nextShrink))
```
I think it reads better as:
```suggestion
                                    if (NanoTime.isAfter(nextShrink, now))
```

sbordet replied:
There is no `NanoTime.isAfter()`, and it was explicitly not added to keep the `NanoTime` API minimal.

gregw replied:
I think it really improves readability, so it is a small price to pay.

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -1027,54 +1057,68 @@ public void run()
-                                    break;
+                                    long lastShrink = _lastShrink.get();
+                                    long prevShrink = lastShrink;
+                                    // If the shrink window is too far in the past,
+                                    // advance it to be one idle timeout before now.
```
I don't like using the word window, as we don't fully embrace that abstraction. how about:
```suggestion
                                    // If the shrink threshold is too far in the past,
                                    // set it to just one idle timeout before now.
```

sbordet replied:
Uhm, it is a window between some time in the past and the current time.
And it is a credit-based scheme like the HTTP/2 flow control window, every expired thread reduces the window until zero, and the passing of time gives it more credit.

gregw replied:
I understand what a window system is and how it is an analogy to the algorithm here.
I just do not think that in this case it adds any clarity by describing it as a window.

Best just to say when the next shrink is allowed to happen.

sbordet replied:
> I just do not think that in this case it adds any clarity by describing it as a window.

But it's not a threshold either, and calling it so it's even more confusing.

> Best just to say when the next shrink is allowed to happen.

The value is completely unrelated to "when the next shrink is allowed to happen", so I cannot say that.

I don't understand why you're so against the term "window", it is perfectly representing what it is: a value that has a fixed maximum, can decrement until it closes down to zero, and when it closes, no more air (err shrink), but can reopen 😃 

We can call it `shrinkCredit` if you want, but seems to me unnecessary wording when "window" is precisely describing it.
"Allowance"? "Quota"?

If we change the math to use a counter instead of a nanoTime, how would you call the counter? 
There would be N "shrinkCredits" and each time a thread is eligible to shrink you would decrement by 1 credit.
As time passes you gain 1 credit for every shrink period that has passed.

gregw replied:
A window has two ends that adjust and can't be allowed to get too far apart.  This just doesn't have that abstraction.  It is kind of similar, but not close enough that the term window introduces confusion rather than clarity.
It only works for you because you have implemented a window with a single counter (window size), but the origin of the term really depends on having two pointers and a window between them that is limited in size.  

It is not a quota, nor allowance, nor credit.   It is a time that is related to now, that moves forward in time and acts as a threshold to determine if an expiry is allowed or skipped.

sbordet commented:
@magibney this is an alternative to your proposal.

It works by maintaining a window of 1 idle timeout, and moving the window every time a thread is eligible for shrinking.

The code reads `_lastShrink` as the start of the window, and if it is too far back in the past, it updates the start of the window to be `now - idleTimeout`.

Then, for every thread eligible for shrinking, the start of window is updated by 1 shrink period, until it is less than `now`. This guarantees that only `idleTimeoutMaxShrinkCount` threads are expired within an idle timeout period.

Say the idle timeout is 10s, and `idleTimeoutMaxShrinkCount=5`.
At time `t` an idle thread is eligible to shrink, so the window start is set to `t-10`.
Since `t-10<t`, the thread is exited and the window moved to `t-8`.
At `t+1` another idle thread is eligible to shrink, and the window start is at `t-8`, the thread is exited and the window moved at `t-6`.
At `t+5` another idle thread is eligible to shrink; the window is at `t-6`, but it's larger than one idle timeout, so it is set at `t-5`, the thread is exited and the window moved at `t-3`.
At `t+6` many threads become idle and eligible to shrink; the window is at `t-3`; one thread is exited and moves the window at `t-1`; another exits and moves to `t+1`; another to `t+3`; another to `t+5`. The next one _would_ move the window at `t+7`, but that is past the current "now" `t+6`, so it is not exited and waits another idle timeout.
And so on.

Graphically the example above is:

```
time t, window start at t-10
event            v
time   -9-------1t------
window ^

time t, window updated to t-8
event            v
time   --8------1t------
window   ^

time t+1, window updated to t-6
event            vv
time   ----6-----t1-----
window     ^

time t+5, window updated to t-3
event            vv   v
time   -------3--t----5-
window        ^

time t+6, 3 window updates to t+5
                       3
event            vv   vv
time   ----------t----56
window                ^
```

When many threads idle at the same time, one will set the window start at `t-10` and the first update at t-8, other 4 will do `+2` updates up to `t`, and the others cannot move the window past `t` and will wait another idle timeout.

@magibney can you try this algorithm for your use case?
We would also appreciate a review of this PR.
Thanks!

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -1027,54 +1057,68 @@ public void run()
-                                        LOG.debug("shrinking {}", QueuedThreadPool.this);
-                                    break;
+                                    long lastShrink = _lastShrink.get();
+                                    long prevShrink = lastShrink;
```
```suggestion
                                    long shrinkThreshold = lastShrink;
```

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -1027,54 +1057,68 @@ public void run()
+                                    long prevShrink = lastShrink;
+                                    // If the shrink window is too far in the past,
+                                    // advance it to be one idle timeout before now.
+                                    if (NanoTime.elapsed(lastShrink, now) > idleTimeoutNanos)
+                                        prevShrink = now - idleTimeoutNanos;
```
Could we just do:
```java
    long shrinkThreshhold = Math.max(lastShrink, now - idleTimeoutNanos);
```
or does nano wrapping make this not work?  It is the semantic we want.

sbordet replied:
Yes, it could wrap, so we can't use `Math.max()`.

gregw replied:
doh!

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -1027,54 +1057,68 @@ public void run()
+                                    if (NanoTime.elapsed(lastShrink, now) > idleTimeoutNanos)
+                                        prevShrink = now - idleTimeoutNanos;
+
+                                    // Add shrink periods until the window is full.
```
I don't like "until" in this comment, implies looping until it is full.
```suggestion
                                    // The next shrink can happen one shrink period after the threshold. 
```

sbordet replied:
But it does not "happen" at that time... `nextShrink` is perhaps badly named (but can't find better) but it is unrelated with _when_ the shrink happens (which is now, not at `nextShrink`).

gregw replied:
Maybe call it `nextShrinkAfter`.

Eitherway, the "until" text doesn't help an it is better to just say when the next shrink can happen.

```suggestion
                                    // Another shrink is not allowed before one shrink period after the threshhold.
```

gregw reviewed (changes requested):
just some naming niggles

gregw committed:
rename variables

Signed-off-by: gregw <gregw@webtide.com>

magibney commented:
Thanks for this, it definitely looks promising!

IIUC some of the logic in this PR is quite similar to the earlier iteration of #9354 ([diff here](https://github.com/eclipse/jetty.project/compare/beee5ffff0..d365f55fec7951adb946cae15d517c07731df714)): updating a backdated timestamp upon shrink. The main difference between the approach of this PR and the current state of #9354 (aside from the fact that the latter refactors out the shrink bookeeping into a separate class) is that the latter tracks pool-level idle state as a "first-class citizen" (more discussion below).

It was pretty easy to port my tests over. Some of them pass, but the most realistic one appears to have some issues, namely `testShrinkToSustainedBaseline`:
```
2023-03-15 14:42:08.613:INFO :oejut.QueuedThreadPoolShrinkTest:main: CONFIG: baselineTargetThreadCount=178, expectStableThreadCount=179, idleTimeoutDecay=326, busy=false, shift=3, seed=2a0b1a2dcd4f11ba
2023-03-15 14:42:11.309:INFO :oejut.QueuedThreadPoolShrinkTest:main: OK threadCount expect=835, actual=999, variance=-1.0
2023-03-15 14:42:11.815:INFO :oejut.QueuedThreadPoolShrinkTest:main: OK threadCount expect=670, actual=999, variance=-1.0
2023-03-15 14:42:12.320:INFO :oejut.QueuedThreadPoolShrinkTest:main: OK threadCount expect=505, actual=999, variance=-1.0
2023-03-15 14:42:12.825:INFO :oejut.QueuedThreadPoolShrinkTest:main: OK threadCount expect=341, actual=999, variance=-1.0
2023-03-15 14:42:13.330:INFO :oejut.QueuedThreadPoolShrinkTest:main: OK threadCount expect=176, actual=999, variance=-1.0
2023-03-15 14:42:13.835:INFO :oejut.QueuedThreadPoolShrinkTest:main: OK threadCount expect=11, actual=999, variance=-1.0
too many consecutive variance faults; expect=-153, actual=999, variance=-1.0 (baselineTargetThreadCount=178, expectStableThreadCount=179, idleTimeoutDecay=326, busy=false, shift=3, seed=2a0b1a2dcd4f11ba)
2023-03-15 14:42:14.342:INFO :oejut.QueuedThreadPoolShrinkTest:main: OK threadCount expect=-153, actual=999, variance=-1.0
```
This test spins up 1k threads, then pings the pool periodically with _n_ concurrent threads (_n_ here being `expectStableThreadCount` -- 179 in this case) and waits for the superfluous threads to die, which they don't appear to do.

I think this is happening because of random assignment of threads; if you have sufficient load to randomly "ping" a thread alive within its `idleJobPoll` window, threads might never check eligibility to expire, even if load could be serviced by many fewer threads. I don't think that's different now than it was before, so maybe not a blocker here. And I think the only way around it is to track pool-level idle capacity as a "first-class citizen", e.g. as done in the `ShrinkManager` timestamp deque in  #9354. The only practical downside of _not_ explicitly tracking pool-level idle capacity is that it's possible for a handful of clients, each with a sustained load of short jobs, to artificially inflate the pool size. I'll take a look at verifying that this behavior already existed (i.e., was not introduced by this PR).

That said, even if the behavior hasn't (strictly speaking) changed, the user expectations _will_ have changed, in a way that may make more apparent the disparity between how quickly the user thinks the pool _should_ shrink (based on shrink rate configuration) and how quickly it _actually_ shrinks.

`testShrinkToSustainedBaseline` is arguably the most realistic test among the suite of tests I introduced in #9354, in the sense that it tests idleTimeout behavior on a pool that's continuing to get active traffic. I added it to ensure that the new approach would work as expected in a production system, and it's designed to apply more stress than one might reasonably expect in a real production deployment. So the question of whether this alternative PR will meet our needs in practice is harder to answer: a passing `testShrinkToSustainedBaseline` strongly suggests that a solution _will_ suffice, but a failing `testShrinkToSustainedBaseline` doesn't necessarily mean it _won't_ suffice ...

magibney commented:
>I'll take a look at verifying that this behavior already existed (i.e., was not introduced by this PR).

I followed up on this, and it turns out that the `testShrinkToSustainedBaseline` test, when adjusted to work with the default shrink rate of 1 (`CONFIG: baselineTargetThreadCount=954, expectStableThreadCount=955, idleTimeoutDecay=1, busy=false, shift=1, seed=9efe9d4bb15bfdc3`) works fine against the `main` branch (shrinks as expected to the `baselineTargetThreadCount` --  works on  #9354 too fwiw), but doesn't shrink at all on this PR branch. So there's definitely something that needs to be addressed there ...

gregw commented:
Can you make the source of your test available, so we can work on the exact same use-case.

gregw committed:
Tried a better shrink algorithm.

Not all tests passing yet.  Problem with idle accounting.

Signed-off-by: gregw <gregw@webtide.com>

gregw commented:
I've pushed a slightly different approach.

I added a test that spikes 1000 threads then keeps 200 busy. Old approach didn't shrink. This new approach does shrink down to 200.

However, there are some problems with idle thread accounting, as some other tests fail.

I'm on vacation from tomorrow, so pushing so @sbordet can find my mistakes.

gregw committed:
Tried a better shrink algorithm.

Fixed the tests

Signed-off-by: gregw <gregw@webtide.com>

gregw commented:
@sbordet I think I have fixed the idle accounting.
However, I had to adjust a test as we kind of have a race when a job finishes and the thread might go idle or might now exit.   Green build for me, but needs careful review.

magibney commented:
Making sure you've seen #9512, opened against this PR. The most recent changes don't actually seem to respect the "peak hold" aspect of `idleTimeout` -- i.e., I think threads start expiring immediately.

Feel free to merge that test suite in here if you like. I also took a bit of extra effort to make it easy to run against `main`, to validate parity of default behavior (should be only a one-line change necessary when porting to `main` branch).

gregw commented:
@magibney Yes I have seen that. Thanks.

Yeah, this PR is not going to wait an idleTimeout before starting to shrink.  It will shrink up to the shrink count threads every idle period, starting immediately

magibney commented:
> Yeah, this PR is not going to wait an idleTimeout before starting to shrink. It will shrink up to the shrink count threads every idle period, starting immediately

Perhaps I'm missing something, but then how does this differ from the behavior of the current `main` branch if you divide `idleTimeout / idleTimeoutMaxShrinkCount`?

gregw commented:
The difference with this latest branch is when we do the shrink check. With previous versions we did the check only when we idle timed out... Which if we had enough small busy tasks, never happened, so 200 concurrent tasks could keep 1000 threads not idle 

We now do the shrink check whenever the queue is empty of jobs. So with 200 busy threads the pool still shrinks from 1000 to 200.

However the shrinking starts immediately after a spike in the threads.  

The only thing we could possibly add is a minimum time to live. So if a thread has been alive for less than an idle timeout, it will not shrink.  This would make the peaks fatter. @sbordet thoughts?

magibney commented:
Oh! I see. Then iiuc the current incarnation of this PR isn't so much a replacement for #9354 as it is a fix for a latent issue surfaced by tests that were added in #9354. In that case though, for the current state of this PR I don't see the purpose of adding the `idleTimeoutMaxShrinkCount` param: if you start to shrink immediately then setting `idleTimeout=X, idleTimeoutMaxShrinkCount=Y` would be functionally equivalent to setting `idleTimeout=X/Y`, no?

gregw commented:
@magibney yeah the idle timeout and count now make less sense... unless we add in a min time to live equal to the idle timeout.  We need to think about this some more, but I like the direction we are going.   However, we need to do some performance testing on this to see if we have changed anything one way or the other.

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -1027,89 +1069,94 @@ public void run()
                     }
                     catch (InterruptedException e)
                     {
+                        boolean interrupted = Thread.interrupted();
```
Won't this virtually always be false as `InterruptedException` is supposed to be thrown after a `Thread.interrupted()` returns true?

lorban reviewed (commented):
Performance wise, I don't think these changes make any difference: a loaded pool isn't going to shrink and spends most of its time in `BlockingArrayQueue`'s `offer` and `poll` methods plus on the `_counts` atomic. The few accesses to the `_shrinkThreshold` atomic only happen when the pool has no queued job, so the extra needed cycles shouldn't matter.

magibney commented:
Thanks for confirming. Yes, the diff for the current iteration could be a lot more straightforward -- basically an internal refactoring of the loop around `idleJobPoll()` (and any related tests), with no documentation/API updates necessary.

Separately considering the "minimum TTL" thing: I'm increasingly thinking that, at a high level, an approach that extracts the shrink logic out into an interface (similar to what's done with `ShrinkManager` in #9354) would be beneficial.

The implementation in #9354 supports "minimum TTL" and then linear decay; but given the interface, one could write a trivial implementation that recreates exactly the current behavior, supporting immediate linear decay via updates to a single `AtomicLong`. It also occurs to me that the `ShrinkManager` impl in #9354 could be straightforwardly tweaked to support exponential decay (asymptotic to `minThreads`)!

I'm not suggesting that any more complex shrink algorithms be implemented in jetty; but simply that the lifecycle hooks necessary to support pluggable decay are quite clear:
```java
/** called to register a thread as idle (not actively performing work) */
void onIdle();

/** called to register a thread as about to perform work */
void onBusy();

/** called to check whether the pool should shrink by at least on thread */
boolean pollIdleShrink();

/** fallback to ensure consistency upon expected thread death */
void prune();
```
Or however one might prefer to rename the methods?

I'm fairly confident btw that introducing the concept of "minimum TTL" will inevitably run into the same problem that motivated the creation of `ShrinkManager` on #9354: we're either tracking TTL baseline per thread (in which case behavior varies arbitrarily according to thread scheduling), or we're tracking via a single pool-level variable (similar to what we do now) in which case it's not clear what TTL applies to. Or, as in the approach taken by `ShrinkManager`, we must track the TTL baseline of all threads individually but interchangeably at the pool level, and act on them in a consistent way.

magibney commented:
> Performance wise, I don't think these changes make any difference

Hmm this makes sense looking at the code for current `main` branch. And as I [mentioned above](#issuecomment-1470833629), the "sustained baseline" test indeed passed on current `main` branch; it only failed on the earlier iteration of this branch. So I suspect that the current iteration of this PR incidentally removed the functionality that motivated this PR while fixing a problem that was introduced in the earlier iteration of this PR.

lorban committed:
#9237 improve comments, interrupt handling and add tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#9237 add QTP benchmark

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw commented:
We have a release coming in the next week or two. We could merge this as is, or work on it a bit more for the next release.

I'm not opposed to making the algorithm extensible.

I'm also cautious about adding configuration that we are not 100% on, as once in it is forever... hence i lean towards skipping this release cycle.

@sbordet thoughts?

lorban committed:
#9237 add extra aggressive shrinking flag to control if the idle timeout period should be observed or not

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

magibney commented:
At a high level, I think this PR in its current state, for config
```properties
idleTimeout=X
idleTimeoutMaxShrinkCount=Y
aggressiveShrinking=true
```
is functionally nearly identical to current `main` branch for config
```properties
idleTimeout=X/Y
```
IIUC both cases should shrink 1 thread per `X/Y` interval.

At this point I'm honestly not sure what problem this PR is trying solve.

I'm pretty sure existing behavior on main branch is equivalent to `aggressiveShrinking=true`, and that setting `aggressiveShrinking=false` would put you at substantial risk of having a pool that never shrinks (by making it possible for a thread to acquire all its jobs via `idleJobPoll()`, without ever checking eligibility to shrink).

I'm still confident in #9354 and happy to answer any questions if you're open to taking another look at it. Even if just to use that as a base for pluggable shrinking (again, extracting the shrink logic, as in the commit I just pushed to that branch, seems like a win regardless of whether the purpose is to make it pluggable, or just to gain some logical separation).

lorban committed:
#9237 make sure protected runJob() always is called within a try-catch block and that the interrupt flag is always cleared when it returns, also make aggressive shrinking the only possibility

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#9237 make shrinkIfNeeded() protected

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
@magibney The `aggressiveShrinking` config param I added was an experiment I made to compare two slightly different design choices. I've now removed it and made what was the aggressive mode the only implementation.

I'm sorry that I haven't been very open about the work I've been doing, but your pull request shed some light on a very interesting problem, and I needed to fully grasp all the details before taking any decision as this is a crucial piece of code. But I'm now done and you can consider the current state of this pull request as the most likely final version of the code, so let me give you some details.

This new shrink implementation is different from the previous one in two ways:
 - The previous implementation requires the thread pool to be virtually idle before it starts shrinking. If it handles a peak that makes the thread count grow to max and after that you always apply a moderate load that keeps just a few threads constantly busy, it will never shrink. The new one configured with `aggressiveShrinking=false` reproduced this behavior, and I added this parameter for the sole purpose of comparing both to better understand them. This new implementation will always try to shrink the pool as soon as one thread runs out of jobs, as we decided it was an improvement worth doing that didn't have any noticeable drawback, especially on the performance side of things.
- You are right that the old implementation with `idleTimeout=X/Y` would shrink at the same pace (assuming your pool becomes idle of course), except for one detail: this one is a lot more performant as it can make Y threads die for each `idleTimeout` period, which basically means the threads needs to be awoken Y times less often by the scheduler before reaching a new stable point.

Thanks for putting us on our toes, we revisited and improved this core component of Jetty in a way we never thought about. Unfortunately, we decided against integrating your suggested shrink manager in the standard code base as its abstraction looks too specific to your implementation, it suffers from a few bugs (the `long[]` is read and updated concurrently without any form of barrier for instance) and fundamentally relies on an algorithm that creates contention, which makes it noticeably slower than our own. You can check the results of the `QTPBenchmark` to get an idea of the perf difference on a loaded pool, it's quite substantial.

But `runJob(Runnable job)` is `protected`, as well as `shrinkIfNeeded()` now, so it seems that you could re-implement you algorithm by subclassing `QueuedThreadPool` and overriding those two methods that provide you the hooks you need should you want to run with it.

lorban commented:
@sbordet do you have anything more to say about this subject? WDYT of this version?

magibney commented:
Ah, this makes sense. Thanks for the detailed explanation, and apologies for prematurely calling out the `aggressiveShrinking`!

In the course of investigating whether this change would include the requisite hooks for the functionality I'm looking for, I realized that the new logic in this PR makes it much cleaner/easier, and I generally really like this change! -- though at the moment I suspect the hooks aren't sufficient to extend via simple subclassing.

I've opened #9532 against this PR branch. It addresses the performance concerns (the QTPBenchmark shows parity, regardless of configuration of `idleTimeout` or `idleTimeoutMaxShrinkCount`), and all the tests added for #9354 pass as well.

> it suffers from a few bugs (the long[] is read and updated concurrently without any form of barrier for instance)

Regarding this, I really think it's a benign race condition. As mentioned [here](https://github.com/eclipse/jetty.project/pull/9354#issuecomment-1468247201):
> It's good to call this out; I did this knowingly and I think it should be ok. For practical purposes, we can lean on the fact that the stored values are effectively relevant within the tolerance of idleTimeout. Writes only happen at the head, and we don't care about reading those values at all, and the exact values that might overwrite each other are a benign race condition because they'll all be well within the idleTimeout tolerance, so we actually don't care about the exact value that gets written. As far as reading values, we only care about tail, and we only start to pay attention to those tail values once their idleTimout is passed. The race condition at tail does happen with failed CaS; but it's benign here too, because we don't care who is checking a value, so long as someone is. The top priority at the tail is to ensure that the array entries remain in order (within the acceptable tolerance of race conditions), so note that we don't actually ever write to the array at tail, we simply update the tail pointer.
>
>Again, the tests bear all this out, and are designed to stress the system pretty hard, asserting very tight tolerances wrt expected shrink behavior.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -90,11 +90,11 @@ public class QueuedThreadPool extends ContainerLifeCycle implements ThreadFactor
      * <dt>Lo</dt><dd>Net idle threads == idle threads - job queue size.  Essentially if positive,
      * this represents the effective number of idle threads, and if negative it represents the
-     * demand for more threads</dd>
+     * demand for more threads, aka the job queue's size.</dd>
```
```suggestion
     * demand for more threads, which is equivalent to the job queue's size.</dd>
```

lorban replied:
Done.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -936,6 +968,82 @@ protected void runJob(Runnable job)
+        }
+    }
+
+    private void doRunJob(Runnable job)
```
Should not this method be in the `Runner` class?

lorban replied:
Done.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -936,6 +968,82 @@ protected void runJob(Runnable job)
+     * <p>Attempts to shrink the thread pool by one thread if {@link #getThreads()} is greater than {@link #getMinThreads()}.</p>
+     * @return true if shrunk, false otherwise.
+     */
+    protected boolean shrinkIfNeeded()
```
Perhaps rename to just `shrink()`? The `boolean` return type implies that may or may not shrink the current thread.
Or perhaps `evict()` because it refers to the current thread? The shrinking of the pool is a consequence of evicting the current thread.

lorban replied:
`evict()` sounds better, going with that.

gregw replied:
Oh sorry. Didn't see this until after I renamed to `shrink`, which I prefer... but I can live with `evict`.

gregw replied:
I've also fixed the xml and ini file that didn't have the rename done in them.  If you change back to `evict` then don't forget to change them also.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -1026,90 +1134,55 @@ public void run()
             {
-                Runnable job = null;
-                while (true)
+                while (_counts.getHi() > Integer.MIN_VALUE)
```
I'd prefer `!=` instead of `>`.

lorban replied:
Done.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -536,6 +537,37 @@ public void setVirtualThreadsExecutor(Executor executor)
+     * @return the maximum number of idle threads to exit in one idle timeout period
+     */
+    @ManagedAttribute("maximum number of idle threads to exit in one idle timeout period")
+    public int getIdleTimeoutMaxShrinkCount()
```
@gregw suggested to rename this just `maxShrinkCount` but maybe `maxEvictCount` is better?

lorban replied:
`maxEvictCount` sounds better, going with that.

gregw replied:
Oops, I renamed evict to shrink before I saw this.  I've also fixed the xml and ini files that did not have this name change done in them.

sbordet commented on tests/jetty-jmh/src/main/java/org/eclipse/jetty/util/thread/jmh/QTPBenchmark.java:
```diff
@@ -0,0 +1,93 @@
+@State(Scope.Benchmark)
+@Warmup(iterations = 10, time = 1000, timeUnit = TimeUnit.MILLISECONDS)
+@Measurement(iterations = 10, time = 1000, timeUnit = TimeUnit.MILLISECONDS)
+public class QTPBenchmark
```
Rename to `QueuedThreadPoolBenchmark`.

lorban replied:
Done.

sbordet reviewed (commented):
@lorban reviewed -- I request changes but I cannot do it through web UI since I'm the author of the PR.

lorban committed:
#9237 handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

hiteshk25 commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -936,6 +968,65 @@ protected void runJob(Runnable job)
+        while (true)
+        {
+            // No jobs available, should we shrink?
+            int threads = getThreads();
```
It may be minor but looks like there is possibility that thread may read stale `getThreads` value as thread which about to exit may not have decrease this value. That can cause number of threads to alive less than min-threads?

gregw replied:
I think there could be a problem.

It used to work when we could only shrink a single thread per period as if the thread count was stale, then the last shrink time would have been pushed out.

If we are shrinking 10 threads per period, then 10 threads could all wake up, see threads=minthreads+1 and all decide to shrink.  They can all advance the shrink threshold.  If they collide whilst doing so, the they will loop and get the threads count again.  But there is a chance they might not collide on updating the shrink threshhold, so we might let too many threads exit.

thinking....

gregw replied:
@lorban @sbordet this is a difficult one to solve, as we have 2 Atomic counters, and any attempt to keep them "atomic" is bound to be racey.

Currently, we make no CaS operation on the number of threads when a worker thread exits.  Instead we just decrement the thread count and optionally the idle count.

If we are shrinking, perhaps we should CaS the thread count decrement with the number of threads used to make the decision to shrink.  If we fail that CaS, then we have to restore the shrink interval (or more precisely deduct a shrink period from whatever value it currently is) and then loop and try again.  If we success the CaS, we have to make sure the runner does not further decrement the thread count when the thread exits.

I can give this ago... but most times when we think we are being clever with two atomics, we are wrong!

gregw replied:
@lorban @sbordet I've implemented this.  It is not too horrid, but hard to convince myself it is not racy.  But I'm thinking the worst case is we don't shrink when we could.

@magibney I'm struggling to see how this can easily be generalised as the shrink() method needs access to two atomics to CaS on them.

lorban replied:
@gregw The race condition between the two atomics that may make us shrink below `minThreads` does exist, but it's quite rate if the threads execute any code at all. You can convince yourself by instrumenting the pool with a `LongAdder` counting the CAS failures & running `QueuedThreadPoolBenchmark`. It shows these counters aren't very contended, if at all.

Sure, it can still happen. But it's going to be very uncommon and the worst case scenario is that the pool can shrink a few threads below `minThreads`. Do we really need to enforce an invariant at the cost of extra complexity, even if it's moderate?

magibney replied:
Some extra context -- I wrote a test that's intended to trigger this, worst-case scenario, pool shrinking to `minThreads=100` from `maxThreads=1000` and `maxThreads=10`, ~~haven't been able to reproduce this at all on my system, even running hundreds of iterations~~ does reproduce the issue, but only transiently -- `ensureThreads()` brings restores the `minThreads` count. ~~I get that it's possible in theory, and might be more likely to happen on a system with a massive number of cores?~~ In practice I was only ever able to read a thread count 2 threads under `minThreads`. But even at that, any real system would of course also not be jumping through hoops to intentionally try to trigger this issue, so ... ; anyway, here's the test:

```java
    @Test
    public void testMinThreads() throws Exception
    {
        QueuedThreadPool tp = new QueuedThreadPool();
        int minThreads = 100;
        tp.setMinThreads(minThreads);
        int maxThreads = 1000;
        tp.setMaxThreads(maxThreads);
        int idleTimeout = 1000;
        tp.setIdleTimeout(idleTimeout);
        tp.setMaxEvictCount(Integer.MAX_VALUE); // very aggressive
        tp.start();
        CountDownLatch cdl1 = new CountDownLatch(maxThreads);
        CountDownLatch cdl2 = new CountDownLatch(maxThreads + 1);
        for (int i = 0; i < maxThreads; i++) {
            tp.execute(() -> {
                try {
                    Thread.sleep(idleTimeout); // to ensure all are immediately eligible to shrink
                    cdl1.countDown();
                    cdl2.countDown();
                    cdl2.await();
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            });
        }
        cdl1.await();
        assertEquals(maxThreads, tp.getThreads());
        cdl2.countDown();
        while (tp.getThreads() > minThreads) {
            // busy-poll, since we expect this to be really fast.
        }
        Thread.sleep(200); // give it time to drop too low, if it's going to do so
        assertEquals(minThreads, tp.getThreads());
    }
```
I have a clear idea of how this would be generalizable for my use case, but for now it's probably easier to consider this PR in isolation, at least once there's a decision on method naming (looks like converging on `shrink()` etc.?) and whether to strictly enforce minThreads ...

magibney replied:
ahhh so this test may have just been picking up the fact that `ensureThreads()` corrected for overshooting `minThreads` floor -- so maybe the test was triggering the situation after all!

gregw committed:
Fix thread count race

Do a proper CaS on the thread count when shrinking.

Signed-off-by: gregw <gregw@webtide.com>

gregw committed:
Merge remote-tracking branch 'origin/jetty-10.0.x' into fix/jetty-10-9237-improve-qtp-shrink

gregw committed:
Fix thread count race

fixed comment

Signed-off-by: gregw <gregw@webtide.com>

hiteshk25 commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -1023,95 +1125,80 @@ public void run()
-                        runJob(job);
-                        if (LOG.isDebugEnabled())
-                            LOG.debug("ran {} in {}", job, QueuedThreadPool.this);
+                        shrunk = shrink();
```
If thread reaches here after doing `one job immediately`, then it might shirk that thread. And that may cause further thread-creation and thread-exit pattern (thrashing)

I think bigger question here is do we want to make sure thread remain idle for `idletimeout` (or keep-alive time) ?

gregw replied:
I think a minimum time to live of one idle timeout will stop any thrashing.

gregw replied:
Actually, I have just re-added the behaviour that when we start a new thread we advance the shrink threshold to now, so no threads will be able to shrink immediately after one has started.

gregw commented:
@olamy what is going on here: https://jenkins.webtide.net/blue/organizations/jenkins/jetty.project/detail/PR-9498/11/pipeline

gregw committed:
Fix thread count race

fixed rename from evict to shrink
readded setting threshold to prevent thrashing

Signed-off-by: gregw <gregw@webtide.com>

gregw committed:
Fixed references to new setter

Signed-off-by: gregw <gregw@webtide.com>

lorban committed:
#9237 remove handling of the race condition allowing to shrink below minThread + comment why that's safe

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#9237 remove handling of the race condition allowing to shrink below minThread + comment why that's safe

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

hiteshk25 commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -936,6 +970,70 @@ protected void runJob(Runnable job)
+        while (true)
+        {
+            // No jobs available, should we shrink?
+            int threads = getThreads();
```
How about retrying (threads-getMinThreads()) times only?

lorban replied:
This would just trade off a hard CPU consumption limit with a possibility of not shrinking the pool by the configured amount of threads. IMHO the current option is preferable in this case.

hiteshk25 replied:
On line #980,  we have a while loop with true.  Instead of true, we can have this condition. That way we can ensure total thread count doesn't go less than min thread count.

sbordet replied:
The `while` loop is not really a loop, unless the CAS fails.
A thread would enter, and try to move the threshold. If the threshold is already in the future it exits; if it was able to move the threshold, it exits.

hiteshk25 replied:
`The while loop is not really a loop, unless the CAS fails.`
This I wanted to cover that when thread retries it should never go below min-threads. 

Hypothetical example:

1. min-thread = 5
2. We can shrink 10 threads in ideltimeout
3. app creates 3 more threads, total threads = 8 now
4.  All 8 threads reach to that loop at same time, and thread count may go below 5. Thus if every thread tries 8-5 = 3 times only, then only 3 thread will succeed.

lorban replied:
The `ensureThreads()` call in the thread's `finally` block would immediately restart any necessary thread to reach `minThread` again.

sbordet replied:
@hiteshk25 I see what you mean now, let me try to code your suggestion.

sbordet replied:
@hiteshk25 I have pushed the changes to implement your suggestion, please re-review.

sbordet replied:
Note that it is still possible that all 8 threads will be evicted, if none of them fails the CAS.
However, we have `ensureThreads()` in the `finally` block to bring the number of threads back to `minThreads` if that happens.

gregw replied:
More over, if that happens, then we will not shrink the pool for an idleTimeout period, so will will not allow threads to die and then be recreated in a quick cycle.

hiteshk25 commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -1026,92 +1124,75 @@ public void run()
-                        runJob(job);
-                        if (LOG.isDebugEnabled())
-                            LOG.debug("ran {} in {}", job, QueuedThreadPool.this);
+                        if (shrink())
```
Just wanted to raise this, if anyone has opinion on this or something already happening.

https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ThreadPoolExecutor.html
```
Keep-alive times
If the pool currently has more than corePoolSize threads, excess threads will be terminated if they have been idle for more than the
keepAliveTime (see getKeepAliveTime(java.util.concurrent.TimeUnit)). This provides a means of reducing resource consumption when the pool
is not being actively used. If the pool becomes more active later, new threads will be constructed. This parameter can also be changed
dynamically using method setKeepAliveTime(long, java.util.concurrent.TimeUnit). Using a value of Long.MAX_VALUE TimeUnit.NANOSECONDS
effectively disables idle threads from ever terminating prior to shut down. By default, the keep-alive policy applies only when there are
more than corePoolSizeThreads. But method allowCoreThreadTimeOut(boolean) can be used to apply this time-out policy to core threads as
well, so long as the keepAliveTime value is non-zero.
```

Basically, shrink the thread if that is idle for `idletimeout`.  That means, if `job = idleJobPoll(idleTimeout);` returns null, then that thread is eligible for shrink. 

On line `1134`, we have while (job != null) loop, which can return job, job, job, null. After this sequence thread becomes eligible for shrink immediately. That I think we want to avoid it. We should wait for `idletimeout` before we shrink that thread.

Please let me know your opinion on that.

lorban replied:
`shrink()` checks the `shrinkThreshold` timestamp and returns `false` when it is in the future. Every call to `shrink()` moves the `shrinkThreshold` forward a bit, and every call to `startThread()` sets it to one `idleTimeout` period in the future.

This basically means that if the pool decides to grow, then it cannot shrink for a duration at least equal to `idleTimeout`.

lorban committed:
#9237 make startThread set shrinkThreshold in the future to guarantee at least one idleTimeout period without shrinking

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#9237 make startThread set shrinkThreshold in the future to guarantee at least one idleTimeout period without shrinking

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -536,6 +537,37 @@ public void setVirtualThreadsExecutor(Executor executor)
+     *
+     * @param shrinkCount the maximum number of idle threads to exit in one idle timeout period
+     */
+    public void setMaxShrinkCount(int shrinkCount)
```
I think I like better "evict", because we evict threads and as a consequence the pool shrinks, so `maxEvictCount` here and in other places.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -830,7 +863,8 @@ protected void startThread()
                 LOG.debug("Starting {}", thread);
             _threads.add(thread);
-            _lastShrink.set(NanoTime.now());
+            // Update the shrink threshold to prevent thrashing
```
```suggestion
            // Update the shrink threshold to prevent thrashing of just started threads.
```

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -936,6 +970,70 @@ protected void runJob(Runnable job)
+     * <p>Attempts to shrink the current thread from the pool if {@link #getThreads()} is greater than {@link #getMinThreads()}.</p>
+     * @return true if the current thread was shrunk, false otherwise.
+     */
+    protected boolean shrink()
```
Rename to `evict()`?

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -936,6 +970,70 @@ protected void runJob(Runnable job)
+                {
+                    // Yes - we have shrunk, so exit.
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("SHRUNK, threshold={}ms in past {}", NanoTime.millisElapsed(threshold, now), QueuedThreadPool.this);
```
Remove the upper case.

```suggestion
                        LOG.debug("Shrunk, threshold={}ms in past {}", NanoTime.millisElapsed(threshold, now), QueuedThreadPool.this);
```

However, if we agree on "evict", these log statements should be changed.

sbordet reviewed (commented):
Doing these myself, as I realize I'm the author of this PR.

sbordet committed:
Renamed shrink to evict in methods, setters, getter and module files.
Updated javadocs and comments to reflect the code.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
@gregw please see whether you like the rename shrink -> evict (I think it's more correct), otherwise I think we are good with the algorithm and ready to merge.

sbordet committed:
Updates after review.

Now the CAS in evict() is only retried threads-minThreads times,
to avoid to shrink the pool below minThreads.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented:
> @gregw please see whether you like the rename shrink -> evict (I think it's more correct), otherwise I think we are good with the algorithm and ready to merge.

I prefer shrink as it is less imperative, i.e we are shrinking to a minimum size. Evict implies that we are kicking out a thread no matter what.

But i don't mind evict enough to hold this up

hiteshk25 reviewed (approved):
LGTM (Thanks Everyone!)

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -1011,23 +1015,28 @@ protected boolean evict()
+                else
+                {
+                    // Some other thread was evicted.
+                    --threadsToEvict;
```
absolutely trivial, but I much prefer `threadsToEvict--`

gregw commented on jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java:
```diff
@@ -977,23 +977,27 @@ protected void runJob(Runnable job)
+        int minThreads = getMinThreads();
+        int threadsToEvict = threads - minThreads;
+
         while (true)
```
why not just:
```
while(threadsToEvict-- > 0)
```
??

sbordet replied:
Because we want to log when there are no more excess threads, and if we put the condition on the `while` becomes more difficult.

gregw commented:
@sbordet oh this just got merged as I was making some trivial extra review comments.  No problem... if I care about it enough, I'll make another PR for those trivial clean ups (I probably don't)

gregw commented:
Thanks @magibney and @hiteshk25 for your suggestions, help, review and patience while we moved slowly on this critical class.

magibney commented:
Wonderful, thanks again for all your work on this!

sbordet commented:
> Thanks @magibney and @hiteshk25 for your suggestions, help, review and patience while we moved slowly on this critical class.

Totally second that! Thank you!

lorban commented:
> > Thanks @magibney and @hiteshk25 for your suggestions, help, review and patience while we moved slowly on this critical class.
> 
> Totally second that! Thank you!

I agree, you kept us on our toes and it was well worth it. Thanks again!
